### ISSUE TITLE
GMM fitting with full covariance crashes unexpectedly unlike sklearn GMM fitting

### ISSUE THREAD

**Issue by MacDaddio** [borchero/pycave#20]
I was trying to fit a GMM on data and kept getting the same error (with variying numbers for the batch and order):

`_LinAlgError: torch.linalg_cholesky: (Batch element 3): The factorization could not be completed because the input is not positive-definite (the leading minor of order 941 is not positive-definite).`

I made a minimal working example to show when this comes up in practice. I compared to sklearn and somehow sklearn is able to avoid this problem. This issue happens both on CPU and GPU. I have PyCave 3.1.3 and sklearn 0.24.2. Do you have any idea what could be the issue?

Minimum working example:

from pycave.bayes import GaussianMixture
import torch
import numpy as np
from sklearn import mixture

#Set seed
seed = 0
np.random.seed(seed)
torch.manual_seed(seed)

#Inputs
n = 5000
p = 2000
k = 10

#Make some non-Gaussian data
X = np.random.randn(n,p)
X = torch.Tensor(X)
X = torch.nn.ReLU()(X-1)

#Fit Sklearn GMM
gmm_sk = mixture.GaussianMixture(n_components=k,
                                 covariance_type='full',
                                 init_params='kmeans')
gmm_sk.fit(X.numpy())

#Fit PyCave GMM
gmm = GaussianMixture(num_components=k, 
                      covariance_type='full', 
                      init_strategy='kmeans')
gmm.fit(X)

**Comment by borchero** [comment]
Hmm, I've seen this issue occur non-deterministically at some times, thanks for the MWE! I'll try to investigate the issue in the coming days but I'd also be happy about any more input ;)

In the meantime, you might get around your issue by increasing `covariance_regularization`.

**Comment by MacDaddio** [comment]
Increasing `covariance_regularization` did indeed fix the problem in my problem! It is strange that such a problem would occur though. This is my initial thought although I am new to this code. Since the data is unimodal, it is possible that the kmeans initialization results in one or more clusters with very few data points. This results in the covariance of that cluster being rank deficient and hence, not positive definite. However, both the scipy and pytorch library for Cholesky decomposition require postive-definiteness (instead of semi-definiteness) to work so I am not sure what the problem is. I will look into it a bit more tomorrow but thank you for the quick response!

**Comment by MacDaddio** [comment]
dfdd

**Comment by MacDaddio** [comment]
Sorry, I accidently closed this issue.

**Comment by borchero** [comment]
Does this issue occur when you do not perform mini-batch training? Also, I would advise to try using double precision (I think you can pass `precision = 64` to the trainer params).

**Comment by MacDaddio** [comment]
If I use the whole dataset of 10,000 points instead of 1,000 mini-batches, I still get the same issue for covariances regularization under 1.0; however, the 1.0 covariance regularization now works. I passed precision=64 to the trainer with no change in behavior.

**Comment by MacDaddio** [comment]
I just realized that if I initialize with 'kmeans' instead of 'kmeans++', it works fine. So maybe there is something weird going on with 'kmeans++'?

**Comment by MacDaddio** [comment]
When I fit GMMs using the kmeans or kmeans++ initializations, I get a non-positive-definiteness error if the covariance regularization is too low. This error comes from the initialization typically and not necessarily from the fitting of the GMM. Can there be two different covariane regularizations? One for the initialization and one for fitting? Because one may want to have a more regularized initialization so that you get a good start but not necessarily have a heavily regularized GMM fit.
